scrapy/utils/misc.py

```diff
@@ -137,17 +137,24 @@ def create_instance(objcls, settings, crawler, *args, **kwargs):
     ``*args`` and ``**kwargs`` are forwarded to the constructors.
 
     Raises ``ValueError`` if both ``settings`` and ``crawler`` are ``None``.
+
+    Raises ``TypeError`` if the resulting instance is ``None`` (e.g. if an
+    extension has not been implemented correctly).
     """
     if settings is None:
         if crawler is None:
             raise ValueError("Specify at least one of settings and crawler.")
         settings = crawler.settings
+    instance = None
     if crawler and hasattr(objcls, 'from_crawler'):
-        return objcls.from_crawler(crawler, *args, **kwargs)
+        instance = objcls.from_crawler(crawler, *args, **kwargs)
     elif hasattr(objcls, 'from_settings'):
-        return objcls.from_settings(settings, *args, **kwargs)
+        instance = objcls.from_settings(settings, *args, **kwargs)
     else:
-        return objcls(*args, **kwargs)
+        instance = objcls(*args, **kwargs)
+    if instance is None:
+        raise TypeError("Instance is None for %s" % objcls)
```

Comment:
```suggestion
    if not isinstance(instance, objcls):
        raise TypeError("%r is not an instance of %r" % (instance, objcls))
```
I wonder if it wouldn't be better to check if `instance` is actually an instance of the class we're working with.

Reply:
Oh yeah, thanks, will change. I'd forgotten python isn't block scoped.

Reply:
I could do this but it makes the unit tests a bit awkward, because in them we are using a MagicMock object as the objcls argument, rather than a class, so the isinstance() check fails.

I tried a few things e.g. the below (doesn't work) but haven't found a way around it yet so just thought I would check you think it's worth it first.

```python
spec_sets = (['__class__'], ['from_settings', '__class__'], ['from_crawler', '__class__'],
             ['from_settings', 'from_crawler', '__class__'])
for specs in spec_sets:
    m = mock.MagicMock(spec_set=specs)
    m.__class__ = mock.MagicMock
```

Reply:
Hm it does seem tricky. Here are my two cents though:
* I'd prefer to adapt the tests to the implementation, not the other way around.
* I still think it'd be good to check for the class. Consider a class with a `from_crawler` class method that returns `3`. 3 is not None, but it's also probably not what Scrapy expects, I think it would be good to fail early in that case.

Sadly, I'm not a mock expert myself, so I don't have a solution at the moment. But I don't see any reason why the tests should remain how they are, we could write actual classes with the corresponding methods. I know it'd be more verbose than with mocks, but as I said, tests should adapt to the implementation IMHO.

Reply:
Great, thanks for your suggestion.